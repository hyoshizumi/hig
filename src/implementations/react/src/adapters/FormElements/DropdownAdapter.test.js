/* eslint-disable react/prop-types */
/**
 Copyright 2016 Autodesk,Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.

 */

import { mount } from 'enzyme';
import * as HIG from 'hig-vanilla';
import React from 'react';

import DropdownComponent from './DropdownAdapter';

const OptionComponent = DropdownComponent.Option;

const inputId = 'text-field-5678';

describe('<DropdownAdapter>', () => {
  function setLabelForInputId(higContainer) {
    // to adjust for the randomly generated id
    const labels = higContainer.querySelectorAll('label');

    [].forEach.call(labels, (label) => {
      label.setAttribute('for', inputId);
    });

    const input = higContainer.querySelector('input');
    input.setAttribute('id', inputId);
  }

  function createHigContext(props) {
    const higContainer = document.createElement('div');

    const higDropdown = new HIG.Dropdown(props);
    higDropdown.mount(higContainer);

    const higOption = new higDropdown.partials.Option({
      label: props.option.label,
      value: props.option.value
    });

    higDropdown.addOption(higOption);

    setLabelForInputId(higContainer);

    return { higContainer, higDropdown, higOption };
  }

  const Context = props => (
    <DropdownComponent
      label={props.label}
      placeholder={props.placeholder}
      instructions={props.instructions}
    >
      <OptionComponent
        label={props.option.label}
        value={props.option.value}
      />
    </DropdownComponent>
  );

  const UpdatedContext = (props) => {
    const { option, ...dropdownProps } = props;
    return (
      <DropdownComponent {...dropdownProps}>
        <OptionComponent {...option} />
      </DropdownComponent>
    );
  };

  const EventListenerContext = props => (
    <DropdownComponent {...props}>
      <OptionComponent {...props} />
    </DropdownComponent>
  );

  it('renders a dropdown with initial props', () => {
    const defaults = {
      label: 'dropdown label',
      placeholder: 'dropdown placeholder',
      instructions: 'instructions for dropdown',
      option: {
        label: 'option label',
        value: 'option value'
      }
    };
    const { higContainer } = createHigContext(defaults);
    const container = document.createElement('div');
    mount(Context(defaults), {
      attachTo: container
    });

    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();
    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a dropdown with updated props', () => {
    const defaults = {
      label: 'dropdown label',
      placeholder: 'dropdown placeholder',
      instructions: 'instructions for dropdown',
      option: {
        label: 'option-label',
        value: 'option-value'
      }
    };

    const { higContainer, higDropdown, higOption } = createHigContext(defaults);
    const container = document.createElement('div');
    const wrapper = mount(<UpdatedContext {...defaults} />, {
      attachTo: container
    });

    setLabelForInputId(container);

    const option = {
      label: 'updated option label',
      value: 'updated option value'
    };

    const nextProps = {
      label: 'updated dropdown label',
      placeholder: 'updated dropdown placeholder',
      instructions: 'updated instructions for dropdown',
      selectedOptionLabel: option.label,
      option: {
        label: option.label,
        value: option.value,
        selected: true
      }
    };

    higDropdown.setLabel(nextProps.label);
    higDropdown.setPlaceholder(nextProps.placeholder);
    higDropdown.setInstructions(nextProps.instructions);
    higDropdown.setSelectedOptionLabel(nextProps.option.label);
    higOption.setLabel(nextProps.option.label);
    higOption.setValue(nextProps.option.value);
    higOption.select();

    wrapper.setProps(nextProps);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();
    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  it('renders a disabled', () => {
    const defaults = {
      label: 'dropdown label',
      placeholder: 'dropdown placeholder',
      instructions: 'instructions for dropdown',
      disabled: true,
      option: { label: 'option-label', value: 'option-value' }
    };

    const { higContainer, higDropdown } = createHigContext(defaults);
    const container = document.createElement('div');
    mount(<UpdatedContext {...defaults} />, {
      attachTo: container
    });

    higDropdown.disable();
    setLabelForInputId(container);

    expect(container.firstElementChild.outerHTML).toMatchSnapshot();
    expect(container.firstElementChild.outerHTML).toEqual(
      higContainer.firstElementChild.outerHTML
    );
  });

  [
    'onBlur',
    'onTargetClick',
    'onClickOutside',
    'onKeypress',
    'onFocus'
  ].forEach((eventName) => {
    it(`sets event listeners for ${eventName} initially`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(EventListenerContext({ [eventName]: spy }), {
        attachTo: container
      });
      setLabelForInputId(container);

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(
        `${eventName}Dispose`
      );
      expect(disposeFunction).toBeDefined();
    });

    it(`sets event listeners for ${eventName} when updated`, () => {
      const spy = jest.fn();
      const container = document.createElement('div');
      const wrapper = mount(EventListenerContext({}), {
        attachTo: container
      });
      setLabelForInputId(container);

      wrapper.setProps({ [eventName]: spy });

      const instance = wrapper.instance().instance;

      const disposeFunction = instance._disposeFunctions.get(
        `${eventName}Dispose`
      );
      expect(disposeFunction).toBeDefined();
    });
  });
});
