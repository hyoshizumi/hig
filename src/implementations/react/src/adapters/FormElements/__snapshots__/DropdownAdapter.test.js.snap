// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownAdapter> renders a disabled 1`] = `
"<div class=\\"hig__dropdown\\">
    <div class=\\"hig__text-field hig__text-field--disabled\\">
    <!--ICON-SPACER-->

    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon\\" for=\\"text-field-5678\\"></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <span class=\\"hig__text-field__extra hig__text-field__extra--dropdown-caret hig__text-field__extra--show\\"><svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret\\"><title>hig-web-light/basic/form-element/caret</title><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"></polyline></svg></span><!--EXTRA-->
            <div class=\\"hig__dropdown__menu-wrapper\\"><div class=\\"hig__dropdown__menu\\"><div class=\\"hig__dropdown__option \\" data-value=\\"option-value\\">
    <span class=\\"hig__option__label\\">option-label</span>
    <div class=\\"hig__dropdown__option__checkmark\\"><div class=\\"hig__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"checkmark-blue-dark\\"><polygon fill=\\"#087AAD\\" points=\\"10.1802137 16 6 12.1129 7.29586623 10.8214 10.1569902 13.4647 16.6790525 7 18 8.2681\\" fill-rule=\\"evenodd\\"></polygon></svg></div></div>
</div></div></div><!--SLOT-->
            <input name=\\"\\" id=\\"text-field-5678\\" class=\\"hig__text-field__input hig__text-field__input--no-value\\" placeholder=\\"dropdown placeholder\\" value=\\"\\" readonly=\\"true\\" disabled=\\"true\\">
            <label for=\\"text-field-5678\\" class=\\"hig__text-field__label hig__text-field__label--visible\\">dropdown label</label>
        </div>
        <p class=\\"hig__text-field__instructions\\">instructions for dropdown</p>
        <!--INSTRUCTIONS-->
        <!--REQUIRED-NOTICE-->
    </div>
</div><!--FIELD-->
</div>"
`;

exports[`<DropdownAdapter> renders a dropdown with initial props 1`] = `
"<div class=\\"hig__dropdown\\">
    <div class=\\"hig__text-field\\">
    <!--ICON-SPACER-->

    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon\\" for=\\"text-field-5678\\"></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <span class=\\"hig__text-field__extra hig__text-field__extra--dropdown-caret hig__text-field__extra--show\\"><svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret\\"><title>hig-web-light/basic/form-element/caret</title><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"></polyline></svg></span><!--EXTRA-->
            <div class=\\"hig__dropdown__menu-wrapper\\"><div class=\\"hig__dropdown__menu\\"><div class=\\"hig__dropdown__option \\" data-value=\\"option value\\">
    <span class=\\"hig__option__label\\">option label</span>
    <div class=\\"hig__dropdown__option__checkmark\\"><div class=\\"hig__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"checkmark-blue-dark\\"><polygon fill=\\"#087AAD\\" points=\\"10.1802137 16 6 12.1129 7.29586623 10.8214 10.1569902 13.4647 16.6790525 7 18 8.2681\\" fill-rule=\\"evenodd\\"></polygon></svg></div></div>
</div></div></div><!--SLOT-->
            <input name=\\"\\" id=\\"text-field-5678\\" class=\\"hig__text-field__input hig__text-field__input--no-value\\" placeholder=\\"dropdown placeholder\\" value=\\"\\" readonly=\\"true\\">
            <label for=\\"text-field-5678\\" class=\\"hig__text-field__label hig__text-field__label--visible\\">dropdown label</label>
        </div>
        <p class=\\"hig__text-field__instructions\\">instructions for dropdown</p>
        <!--INSTRUCTIONS-->
        <!--REQUIRED-NOTICE-->
    </div>
</div><!--FIELD-->
</div>"
`;

exports[`<DropdownAdapter> renders a dropdown with updated props 1`] = `
"<div class=\\"hig__dropdown\\">
    <div class=\\"hig__text-field\\">
    <!--ICON-SPACER-->

    <div class=\\"hig__text-field__content\\">
        <div class=\\"hig__text-field__input-wrapper\\">
            <label class=\\"hig__text-field__icon\\" for=\\"text-field-5678\\"></label>
            <button class=\\"hig__text-field__extra hig__text-field__password-reveal-button\\"></button>
            <button class=\\"hig__text-field__extra hig__text-field__password-hide-button\\"></button>
            <span class=\\"hig__text-field__extra hig__text-field__extra--dropdown-caret hig__text-field__extra--show\\"><svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret\\"><title>hig-web-light/basic/form-element/caret</title><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"></polyline></svg></span><!--EXTRA-->
            <div class=\\"hig__dropdown__menu-wrapper\\"><div class=\\"hig__dropdown__menu\\"><div class=\\"hig__dropdown__option hig__dropdown__option--selected\\" data-value=\\"updated option value\\">
    <span class=\\"hig__option__label\\">updated option label</span>
    <div class=\\"hig__dropdown__option__checkmark\\"><div class=\\"hig__icon\\"><svg width=\\"24\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"checkmark-blue-dark\\"><polygon fill=\\"#087AAD\\" points=\\"10.1802137 16 6 12.1129 7.29586623 10.8214 10.1569902 13.4647 16.6790525 7 18 8.2681\\" fill-rule=\\"evenodd\\"></polygon></svg></div></div>
</div></div></div><!--SLOT-->
            <input name=\\"\\" id=\\"text-field-5678\\" class=\\"hig__text-field__input\\" placeholder=\\"updated dropdown placeholder\\" value=\\"updated option label\\" readonly=\\"true\\">
            <label for=\\"text-field-5678\\" class=\\"hig__text-field__label hig__text-field__label--visible\\">updated dropdown label</label>
        </div>
        <p class=\\"hig__text-field__instructions\\">updated instructions for dropdown</p>
        <!--INSTRUCTIONS-->
        <!--REQUIRED-NOTICE-->
    </div>
</div><!--FIELD-->
</div>"
`;
